fem/fespace.cpp

```diff
@@ -855,6 +870,181 @@ const Operator *FiniteElementSpace::GetElementRestriction(
    return L2E_nat.Ptr();
 }
 
+const Operator *FiniteElementSpace::GetFaceRestriction(
+   ElementDofOrdering e_ordering, FaceType type) const
+{
+   // Check if we have a discontinuous space using the FE collection:
+   const L2_FECollection *dg_space = dynamic_cast<const L2_FECollection*>(fec);
+   if (dg_space)
+   {
+      if (e_ordering == ElementDofOrdering::LEXICOGRAPHIC)
+      {
+         if (type==FaceType::Interior)
+         {
+            if (L2FI_lex.Ptr() == NULL)
+            {
+               L2FI_lex.Reset(new L2FaceRestriction(*this, e_ordering, type));
+            }
+            return L2FI_lex.Ptr();
+         }
+         else //Boundary
+         {
+            if (L2FB_lex.Ptr() == NULL)
+            {
+               L2FB_lex.Reset(new L2FaceRestriction(*this, e_ordering, type));
+            }
+            return L2FB_lex.Ptr();
+         }
+      }
+      // e_ordering == ElementDofOrdering::NATIVE
+      if (type==FaceType::Interior)
+      {
+         if (L2FI_nat.Ptr() == NULL)
+         {
+            L2FI_nat.Reset(new L2FaceRestriction(*this, e_ordering, type));
+         }
+         return L2FI_nat.Ptr();
+      }
+      else //Boundary
+      {
+         if (L2FB_nat.Ptr() == NULL)
+         {
+            L2FB_nat.Reset(new L2FaceRestriction(*this, e_ordering, type));
+         }
+         return L2FB_nat.Ptr();
+      }
+   }
+   else
+   {
+      if (e_ordering == ElementDofOrdering::LEXICOGRAPHIC)
+      {
+         if (type==FaceType::Interior)
+         {
+            if (L2FI_lex.Ptr() == NULL)
+            {
+               L2FI_lex.Reset(new H1FaceRestriction(*this, e_ordering, type));
+            }
+            return L2FI_lex.Ptr();
+         }
+         else //Boundary
+         {
+            if (L2FB_lex.Ptr() == NULL)
+            {
+               L2FB_lex.Reset(new H1FaceRestriction(*this, e_ordering, type));
+            }
+            return L2FB_lex.Ptr();
+         }
+      }
+      // e_ordering == ElementDofOrdering::NATIVE
+      if (type==FaceType::Interior)
+      {
+         if (L2FI_nat.Ptr() == NULL)
+         {
+            L2FI_nat.Reset(new H1FaceRestriction(*this, e_ordering, type));
+         }
+         return L2FI_nat.Ptr();
+      }
+      else //Boundary
+      {
+         if (L2FB_nat.Ptr() == NULL)
+         {
+            L2FB_nat.Reset(new H1FaceRestriction(*this, e_ordering, type));
+         }
+         return L2FB_nat.Ptr();
+      }
+   }
+}
+
+const Operator *FiniteElementSpace::GetSingleValuedFaceRestriction(
+   ElementDofOrdering e_ordering, FaceType type) const
```

Comment:
Can some code duplication be reduced here by combining with `GetFaceRestriction`?

Reply:
I simplified the code and removed the `GetSingleValuedFaceRestriction` function.